### ISSUE TITLE
OCR to clipboard hook for selections

### ISSUE THREAD

**Issue by kaihendry** [lupoDharkael/flameshot#702]
Sometimes I'm working with screenshots and would be awesome to make a selection around some text and send it off to say https://aws.amazon.com/rekognition/ for it analyse and return the text to set my clipboard.

**Comment by MyriaCore** [comment]
OCR would be great honestly!

**Comment by nsa** [comment]
There is a powerful OCR engine from Google called https://github.com/tesseract-ocr/tesseract and it's open-sourced. It would be super cool if the maintainers could integrate tesseract to flameshot as the author suggested, by doing so there would no need for external API connections such as Amazon's, because tesseract works offline.

**Comment by JScriber** [comment]
There's a project called [dpscreenocr](https://github.com/danpla/dpscreenocr) that does the job, yet the selection is not as convenient and user-friendly as flameshot's.
OCR integrated to flameshot with a dpscreenocr inspired configuration would be really great.

**Comment by enoy19** [comment]
I implemented that feature and created a pull request: #1239

**Comment by kaihendry** [comment]
https://news.ycombinator.com/item?id=27242392 discusses the topic in a MacOS context

**Comment by Pheggas** [comment]
Anything new around this feature request? It would be super cool to me tho. Oftentimes i'm working with text copying and this 
feature would be absolute killer!

**Comment by xeaon** [comment]
@Pheggas If you are comfortable with compiling yourself, check my comment above.

I'm using this feature since then and it works great.

**Comment by borgmanJeremy** [comment]
I am okay to merge this feature if we can:
- Fix the merge request to it passes CI
- Explain how to package / distribute the language packs on all platforms.

**Comment by Pheggas** [comment]
> @Pheggas If you are comfortable with compiling yourself, check my comment above.
> 
> I'm using this feature since then and it works great.

I'd like to. But it would be my first compilation. I'll try to look at it. In case of failure, I'll write you 🙂

**Comment by leoneivaw** [comment]
I am anxious waiting for this feature!
Thanks!

**Comment by irfan798** [comment]
For anybody waiting for the functionality I found this blog post:
https://slint.github.io/blog/ocr-screenshot.html

If you install _tesseract_, it is basically a one liner
```bash
flameshot gui --raw | tesseract stdin stdout | xclip -in -selection clipboard
```

**Comment by ek1ng** [comment]
It's an amazing solution! Thanks!

**Comment by ek1ng** [comment]
> For anybody waiting for the functionality I found this blog post: https://slint.github.io/blog/ocr-screenshot.html
> 
> If you install _tesseract_, it is basically a one liner
> 
> ```shell
> flameshot gui --raw | tesseract stdin stdout | xclip -in -selection clipboard
> ```

It's an amazing solution! Thanks!

**Comment by siva-sub** [comment]
> For anybody waiting for the functionality I found this blog post: https://slint.github.io/blog/ocr-screenshot.html
> 
> If you install _tesseract_, it is basically a one liner
> 
> ```shell
> flameshot gui --raw | tesseract stdin stdout | xclip -in -selection clipboard
> ```

For those using wayland use this command instead:

`flameshot gui --raw | tesseract stdin stdout | wl-copy`

 it requires the installation of [wl-clipboard](https://github.com/bugaevc/wl-clipboard), together with tesseract and the respective tesseract language data package.

For me in arch linux the packages are :
[tesseract](https://archlinux.org/packages/community/x86_64/tesseract/)
[tesseract-data-eng](https://archlinux.org/packages/community/any/tesseract-data-eng/)
[wl-clipboard](https://archlinux.org/packages/community/x86_64/wl-clipboard/)

in arch linux it can be installed with:
`sudo pacman -S tesseract tesseract-data-eng wl-clipboard`

**Comment by xeaon** [comment]
Last year, I've spent a view hours to adapt @enoy19's pull request #1239 to the newer codebase but rm -rf'ed the directory by accident.

Yesterday I've done it again. You can find my fork here: [xeaon/flameshot](https://github.com/xeaon/flameshot). After taking a screenshot, you can choose the OCR tool to copy the tesseract output to your clipboard. If successful there is a notification.

As of now, you need to install the dependency tesseract-ocr and leptonica manually. It won't build in CI but I'm trying to fix it for Linux builds.

As mentioned before, I'm no C++/Qt dev, so this is pretty challenging. Feel free to contribute to my fork, if you have any skills. This feature can be merged, if we can

- pass CI
- Explain how to package / distribute the language packs on all platforms

as @borgmanJeremy mentioned earlier.

**Comment by pitfiend** [comment]
recently  there's a new kid on the block for OCR, https://github.com/TheJoeFin/Text-Grab, it's open source and seems to be a very nice solution. maybe it could be added as a feature.

**Comment by mmahmoudian** [comment]
@pitfiend Thanks for your comment, but I am going to hide your comment, and here is few points to clarify my reasoning:
- The software you suggested it platform specific (Windows)
- If we implement a feature, our emphasis would be to implement it in a cross-platform way. There are some features that are Linux and macOS specific (like support of CLI), but Windows will never be our main focus, unless the maintainers of the project collectively agree otherwise. For now, the best-case scenario for the C# software you suggested is via a Windows-specific plugin.
- The OCR can easily be implemented with 3 lines of shell script (detailed explanation here: https://mehrad.ai/posts/20210702-extracting-payment-info-in-rasterized-invoice/#ocr ).

**Comment by Nasreddine** [comment]
This is my version in Ubuntu using custom shortcuts.

I've created a shell script named `shot.sh` and placed it in `/home/me/apps/`.  [change **me** to your username]
The script contains the following line:

```bash
flameshot gui --raw | tesseract stdin stdout | tr -d '\n' | xclip -in -selection clipboard
```

The addition of **`tr -d '\n'`** is specifically to remove line breaks from the output.

To utilize this script in Ubuntu, you can set up a custom keyboard shortcut as follows:

1. Open **Settings**.
2. Navigate to **Keyboard** > **Keyboard Shortcuts**.
3. Click on **View and Customize Shortcuts**.
4. Under **Custom Shortcuts**, click the **+** sign to add a new shortcut.
5. In the new shortcut setup, enter the following details:
   - **Name**: `flameshot_tesseract`
   - **Command**: `sh /home/me/apps/shot.sh`
   - **Shortcut**: [Choose your preferred key combination]

This will enable you to use the script conveniently with a keyboard shortcut of your choice.

Linux makes digital life easier : )

**Comment by Sheldonimo** [comment]
Hi guys, this is my custom version to linux mint.

I have been exploring methods to enhance the OCR output, and my first step was to define the language settings, which can also involve combinations. Through varying the `--psm` parameters in Tesseract, I discovered that the value of 6, corresponding to "Assume a single uniform block of text," yields the most favorable results. Below is the complete command for your experimentation:
`flameshot gui --raw | tesseract stdin stdout -l eng+spa --psm 6 --oem 3 | xclip -in -selection clipboard`
It's noteworthy that the optimal performance is achieved using Tesseract version 5. There are various ways to download or even compile it. I'm attaching the simplest method, which is through the PPA:
`sudo add-apt-repository -y ppa:alex-p/tesseract-ocr-devel`

This PPA was sourced from the following link, which originates from the official Tesseract page:
- https://notesalexp.org/tesseract-ocr/#tesseract_5.x
- https://tesseract-ocr.github.io/tessdoc/#5xx

**Comment by erfanium** [comment]
+1 tesseract ocr should be a built-in functionality

**Comment by GrabbenD** [comment]
> +1 tesseract ocr should be a built-in functionality

Here's a really good PR :tada: 
https://github.com/flameshot-org/flameshot/pull/3074

It checks for Tesseract in PATH meaning the feature is opt-in and uses upstream updates rather than a statically linked library _(which would inevitably become outdated)_

**Comment by X88R88** [comment]
How do i run flameshot with tesseract on windows?

**Comment by f-x1-1** [comment]
We need built-in OCR. Coming from someone familiar with ShareX and CleanShotX, I can see FlameShot has the potential to get there. Features like OCR, GIF recording, and video capture would make FlameShot a better tool. As someone new to Linux, I haven't found any screen capture tool that even comes close to ShareX and CleanShotX.

**Comment by mmahmoudian** [comment]
@f-x1-1 
> Features like OCR, GIF recording, and video capture would make FlameShot a better tool

As it have been discussed, the dev team unanimously agreed that none of these is in the scope of a screenshot tool. We will only and only allow these via the plugin system.

@X88R88 
> How do i run flameshot with tesseract on windows?

I have provided some explanation here:
https://github.com/flameshot-org/flameshot/issues/702#issuecomment-1303664396

**Comment by f-x1-1** [comment]
> @f-x1-1 
> > Features like OCR, GIF recording, and video capture would make FlameShot a better tool
> 
> As it have been discussed, the dev team unanimously agreed that none of these is in the scope of a screenshot tool. We will only and only allow these via the plugin system.
> 
> @X88R88 
> > How do i run flameshot with tesseract on windows?
> 
> I have provided some explanation here:
> https://github.com/flameshot-org/flameshot/issues/702#issuecomment-1303664396


This sounds good. Similar to what Flow Launcher has come up with. Instead of implementing features natively, they have implemented a plugin system. That way, the app itself isn't bloated with unnecessary features. You choose what you want to install. (It's a universal search engine)

**Comment by GrabbenD** [comment]
> I'm not sure if such a complicated behavior could be implemented as a plugin

@emvaized For reference, this was discussed here: https://github.com/flameshot-org/flameshot/pull/3074#issuecomment-2081471217

**Comment by immanuelfodor** [comment]
OCR text from selection on screen is now basically a standard feature on Samsung Galaxy phones with One UI.
I see it as a must have for desktop, too, and preferably with Flameshot already offering the selection area feature 🚀 

Edit: screenshot of OCR from screenshot on Android

<img width="1439" height="905" alt="Image" src="https://github.com/user-attachments/assets/e48dac9f-916f-408d-aafc-8c37c5180ffb" />

**Comment by coolrecep** [comment]
I just switched from Windows 11, which has OCR feature as default and found this thread. Why did it take 5 years and still no result despite the pull requests?

**Comment by hunter-gatherer8** [comment]
Could somebody please clarify, what's the current stance on this? As @mmahmoudian mentioned in #4228 , the maintainers' position is that this should be a plugin (for pretty obvious reasons, to be fair), but I failed to find any information about plugins.

Of course, all of this can be done in 10 lines of bash with `flameshot gui --raw` and subsequent call to `tesseract`, but that means you have to bind a special hotkey via Gnome/Xfce/Qt/..., and it would be far more convenient to just have a button/shortcut in Flameshot, that will call `your-script.sh` with that screenshot (temp file or whatever). Makes sense, especially since you could similarly easy implement google search, LLM integration or basically anything you want.

But it doesn't seem to be currently possible, is it? So, what is that "plugins" system @mmahmoudian was talking about? Is it something in the works, but not ready yet, or an abstract idea with no concrete plans of execution so far?

**Comment by mmahmoudian** [comment]
@hunter-gatherer8 this is very much possible with bash as I have posted this years ago [on my blog](https://mehrad.ai/posts/20210702-extracting-payment-info-in-rasterized-invoice/#ocr)

Regarding the plugin system, there is an open PR for it, but it is not nearly complete:
https://github.com/flameshot-org/flameshot/pull/3661




The following two PRs have tried to implement this into Flameshot, but due to being out of the scope of a screenshot app, they are not merged:

- https://github.com/flameshot-org/flameshot/pull/3074
- https://github.com/flameshot-org/flameshot/pull/4228

**Comment by hunter-gatherer8** [comment]
@mmahmoudian thanks for the answer. By "not possible" I meant "not possible to bind a hotkey **within** Flameshot to execute `$CMD <filename>` after the screenshot is taken". In the blog you seem to only address CLI usage, which is obviously possible. The point is that if you want to bind that script to a hotkey, you now have to do it via Gnome/Xfce/Qt/..., which is inconvenient, because there are only so many keys on your keyboard, so it would be far more convenient (and doesn't seem that complicated to implement) to allow binding a hotkey to a `save + <exec arbitrary command>` action in the Flameshot GUI itself.

**Comment by mmahmoudian** [comment]
@hunter-gatherer8 
> The point is that if you want to bind that script to a hotkey, you now have to do it via Gnome/Xfce/Qt/..., which is inconvenient, because there are only so many keys on your keyboard

Yes, the window manager should coordinate running that script. I personally I don't think there is shortage of keybinding options. You can mix print screen with any or all combinations of Super (Windows key), Shift, Ctrl, Alt. So technically you have 2^4=16 options  with print screen button.

> it would be far more convenient (and doesn't seem that complicated to implement) to allow binding a hotkey to a save + <exec arbitrary command> action in the Flameshot itself

Well, Flameshot should only expose certain things to the plugin, and avoid running external arbitrary commands for users' safety. Also there are a lot of things to consider while designing a plugin system. You can read more about it in:

- https://github.com/flameshot-org/flameshot/discussions/2529
- #3661
- https://github.com/flameshot-org/flameshot/discussions/3953

**Comment by coolrecep** [comment]
> [@hunter-gatherer8](https://github.com/hunter-gatherer8) this is very much possible with bash as I have posted this years ago [on my blog](https://mehrad.ai/posts/20210702-extracting-payment-info-in-rasterized-invoice/#ocr)
> 
> Regarding the plugin system, there is an open PR for it, but it is not nearly complete: [#3661](https://github.com/flameshot-org/flameshot/pull/3661)
> 
> The following two PRs have tried to implement this into Flameshot, but due to being out of the scope of a screenshot app, they are not merged:
> 
> * [Add OCR acction with tessercat #3074](https://github.com/flameshot-org/flameshot/pull/3074)
> * [OCR Tesseract integration  #4228](https://github.com/flameshot-org/flameshot/pull/4228)

I don't get the "being out of the scope of a screenshot app" part. There are screenshot apps with OCR feature:

### **Windows**

* ShareX
* PowerToys (Text Extractor)
* Snipping Tool
* Text Grab

### **macOS**

* CleanShot X
* Shottr
* TextSniper
* TRex
* OwlOCR

### **Linux**

* NormCap
* Frog
* TextSnatcher
* GImageReader
* Spectacle (KDE - newer versions)

**Comment by hunter-gatherer8** [comment]
> Flameshot should only expose certain things to the plugin, and avoid running external arbitrary commands for users' safety

@mmahmoudian  That is true if we are talking about *plugins* that may have been implemented by a 3rd party. However, if we are talking about an arbitrary command that a user assigns himself to a hotkey within Flameshot, there's nothing you should be protecting a user from. In fact, this is the standard approach, and all sorts of programs (in fact, even XFCE's default screenshot tool) allow you to assign an arbitrary command to execute. Actually, setting a hotkey in the window manager also is nothing more than just allowing a user to enter an arbitrary command to execute.

And, obviously, it's not for me to judge, but exactly because of lack of all these intricacies, this seems to be a **much** simpler feature than full-featured plugin API. Maybe we should open a separate feature request, specifically for that? That is, separate from API and/or OCR discussions. (I guess, 90% of use-cases for a more complicated API would be already solved by that relatively simple feature.)

**Comment by mmahmoudian** [comment]
@hunter-gatherer8 
> Maybe we should open a separate feature request, specifically for that?

Sure. Feel free to open a feature request. I would just ask you to consider how this can be done in a cross platform way, and also suggest where/how such scripts should be stored.

Suggestions are always welcome 🤗

**Comment by Rullix** [comment]
In case anyone wants an OCR button, I made my own solution and I'm sharing it [here](https://github.com/Rullix/flameshot-with-ocr-button/releases)

<img width="897" height="666" alt="Image" src="https://github.com/user-attachments/assets/c2c3e0f9-81fb-48ae-a753-51631d5e54a4" />

**Comment by mmahmoudian** [comment]
We have **zero-tolerance** policy for bad mouthing and being nasty. The first instance of uncivilized sentence equals immediate **permanent ban**. Working on Flameshot might be a charity work, but our nerves are definitely not included in this charitable volunteer work.
